mkdocs/utils/__init__.py

```diff
@@ -479,3 +479,47 @@ def nest_paths(paths):
         branch.append(path)
 
     return nested
+
+def get_edit_url(url, path):
+    """
+    Returns a full URL by combining url and path, similarly to urlparse.urljoin(), but
+    without ignoring the search query in the base url.
+
+    This works by moving the ignored query of the first parameter to the second parameter
+    where the urljoin() does preserve the query. Example:
+
+    urljoin('http://base/query?path=', 'docs/index.md')
+    >>> http://base/docs/index.md
+
+    get_edit_url('http://base/query?path=', 'docs/index.md')
+    # Moves the query to the second parameter
+    # Calls: urljoin('http://base/query', '?path=docs/index.md')
+    >>> http://base/query?path=docs/index.md
+    """
+    url_parsed = urlparse(url)
+    path_parsed = urlparse(path)
+
+    # Build the new path_parsed: <url_query> + <path_base> + [path_query]
+    if url_parsed.query:
+        # Setup <url_query> for concatenation
+        url_query = url_parsed.query
+        if not url_query.endswith('/'):
+            url_query += '/'
+
+        if not path_parsed.query:
+            path_parsed = path_parsed._replace(
+                query=url_query + path_parsed.path,
+                path='')
+        else:
+            path_string = path_parsed.path
+            if not path_string.endswith('/'):
+                path_string += '/'
+            path_parsed = path_parsed._replace(
+                query=url_query + path_string + path_parsed.query,
+                path='')
+
+    # new_path contains the query string that will be ignored from new_url
+    new_url = urlunparse(url_parsed)
+    new_path = urlunparse(path_parsed)
+
+    return urljoin(new_url, new_path)
```

Comment:
This seems like an awful lot of code just to support query string URLs. Is this all really necessary? I'm not even convinced we need to use `urljoin`. If we do, fine, but if not then removing that use seems like a better option.

Also, if we explicitly support query strings, then we should explicitly support hash urls as well (`http://example.com/path/to/root/#/foo/bar/page.md`).